File: modules/neomovies-core/ios/AllohaRuntimeParser.swift
import Foundation

enum AllohaRuntimeParser {
    static func parsePayload(_ payload: String, baseURL: String, headers: [String: String]) -> [String: Any]? {
        guard let url = URL(string: baseURL) else { return nil }

        if let stream = parseAllohaBNsiStream(payload, baseURL: url, headers: headers) {
            return stream
        }

        if let fallback = firstPreferredStreamURL(in: payload, baseURL: url) {
            return [
                "videoURL": fallback.absoluteString,
                "audioTracks": [],
                "audioVariants": [],
                "subtitles": subtitleTracks(in: payload, baseURL: url),
                "qualityVariants": [],
                "httpHeaders": headers
            ]
        }

        return nil
    }

    private static func parseAllohaBNsiStream(_ payload: String, baseURL: URL, headers: [String: String]) -> [String: Any]? {
        let candidates = [payload] + embeddedJSONObjectCandidates(in: payload)

        for candidate in candidates {
            guard let data = candidate.data(using: .utf8),
                  let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let source = object["hlsSource"] as? [[String: Any]] else {
                continue
            }

            var qualityVariants: [[String: Any]] = []
            var audioVariants: [[String: Any]] = []
            var masterURL: URL?

            for (index, item) in source.enumerated() {
                guard let quality = item["quality"] as? [String: Any] else { continue }
                var itemVariants: [[String: Any]] = []
                var itemMasterURL: URL?

                for (label, rawValue) in quality {
                    for rawURL in qualityURLStrings(from: rawValue) {
                        let urls = allohaURLs(from: rawURL, baseURL: baseURL)
                        if masterURL == nil {
                            masterURL = urls.first(where: { $0.lastPathComponent.lowercased().contains("master.m3u8") })
                        }
                        if itemMasterURL == nil {
                            itemMasterURL = urls.first(where: { $0.lastPathComponent.lowercased().contains("master.m3u8") })
                        }
                        guard let target = urls.first(where: { !$0.lastPathComponent.lowercased().contains("master.m3u8") }) ?? urls.first else {
                            continue
                        }

                        let variant: [String: Any] = [
                            "label": normalizedQualityLabel(label),
                            "bandwidth": NSNull(),
                            "resolution": NSNull(),
                            "url": target.absoluteString
                        ]
                        itemVariants.append(variant)
                        qualityVariants.append(variant)
                    }
                }

                let chosenAudioURL = itemMasterURL ?? itemVariants.last.flatMap { URL(string: ($0["url"] as? String) ?? "") }
                if let chosenAudioURL {
                    audioVariants.append([
                        "id": "\(index)-\(chosenAudioURL.absoluteString)",
                        "title": audioVariantTitle(from: item, index: index),
                        "url": chosenAudioURL.absoluteString,
                        "qualityVariants": itemVariants
                    ])
                }
            }

            let deduplicatedAudioVariants = deduplicatedAudioVariants(audioVariants)
            let firstURL = deduplicatedAudioVariants.first?["url"] as? String
            let pickedURL = firstURL ?? masterURL?.absoluteString ?? (qualityVariants.last?["url"] as? String)
            guard let pickedURL else { continue }

            return [
                "videoURL": pickedURL,
                "audioTracks": [],
                "audioVariants": deduplicatedAudioVariants,
                "subtitles": subtitleTracks(in: payload, baseURL: baseURL),
                "qualityVariants": qualityVariants,
                "httpHeaders": headers
            ]
        }

        return nil
    }

    private static func deduplicatedAudioVariants(_ variants: [[String: Any]]) -> [[String: Any]] {
        var seen = Set<String>()
        return variants.filter { variant in
            let key = variant["url"] as? String ?? ""
            return seen.insert(key).inserted
        }
    }

    private static func audioVariantTitle(from item: [String: Any], index: Int) -> String {
        if let title = (item["translation"] as? [String: Any])?["name"] as? String, !title.isEmpty {
            return title
        }
        if let title = item["title"] as? String, !title.isEmpty {
            return title
        }
        return "Озвучка \(index + 1)"
    }

    private static func qualityURLStrings(from value: Any) -> [String] {
        if let string = value as? String {
            return string
                .split(separator: ",")
                .map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
        }
        if let values = value as? [Any] {
            return values.flatMap { qualityURLStrings(from: $0) }
        }
        return []
    }

    private static func allohaURLs(from raw: String, baseURL: URL) -> [URL] {
        let decoded = raw.replacingOccurrences(of: "\\/", with: "/")
        let parts = decoded.components(separatedBy: " or ")
        return parts.compactMap { part in
            let clean = part.trimmingCharacters(in: .whitespacesAndNewlines)
            if clean.hasPrefix("//") {
                return URL(string: "https:\(clean)")
            }
            return URL(string: clean, relativeTo: baseURL)?.absoluteURL
        }
    }

    private static func normalizedQualityLabel(_ label: String) -> String {
        let clean = label.replacingOccurrences(of: "_", with: " ").trimmingCharacters(in: .whitespacesAndNewlines)
        return clean.isEmpty ? "Auto" : clean
    }

    private static func firstPreferredStreamURL(in payload: String, baseURL: URL) -> URL? {
        let pattern = #"https?:\\/\\/[^\"'\s>]+\\.(m3u8|mpd)[^\"'\s>]*"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else { return nil }
        let range = NSRange(payload.startIndex..<payload.endIndex, in: payload)
        guard let match = regex.firstMatch(in: payload, options: [], range: range),
              let matchRange = Range(match.range(at: 0), in: payload) else {
            return nil
        }
        let value = String(payload[matchRange]).replacingOccurrences(of: "\\/", with: "/")
        return URL(string: value, relativeTo: baseURL)?.absoluteURL
    }

    private static func subtitleTracks(in payload: String, baseURL: URL) -> [[String: Any]] {
        let pattern = #"https?:\\/\\/[^\"'\s>]+\\.(vtt|srt)[^\"'\s>]*"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else { return [] }
        let range = NSRange(payload.startIndex..<payload.endIndex, in: payload)
        return regex.matches(in: payload, options: [], range: range).compactMap { match in
            guard let matchRange = Range(match.range(at: 0), in: payload) else { return nil }
            let value = String(payload[matchRange]).replacingOccurrences(of: "\\/", with: "/")
            guard let url = URL(string: value, relativeTo: baseURL)?.absoluteURL else { return nil }
            return ["url": url.absoluteString, "label": "Subtitle", "language": "ru"]
        }
    }

    private static func embeddedJSONObjectCandidates(in payload: String) -> [String] {
        var candidates = balancedJSONObjectCandidates(containing: #"\"hlsSource\""#, in: payload)
        candidates.append(contentsOf: balancedJSONObjectCandidates(containing: "hlsSource", in: payload))
        var seen = Set<String>()
        return candidates.filter { seen.insert($0).inserted }
    }

    private static func balancedJSONObjectCandidates(containing marker: String, in payload: String) -> [String] {
        var candidates: [String] = []
        var searchStart = payload.startIndex
        while let markerRange = payload.range(of: marker, options: [.caseInsensitive], range: searchStart..<payload.endIndex) {
            guard let objectStart = payload[..<markerRange.lowerBound].lastIndex(of: "{"),
                  let objectEnd = balancedObjectEnd(from: objectStart, in: payload) else {
                searchStart = markerRange.upperBound
                continue
            }
            candidates.append(String(payload[objectStart...objectEnd]))
            searchStart = markerRange.upperBound
        }
        return candidates
    }

    private static func balancedObjectEnd(from start: String.Index, in payload: String) -> String.Index? {
        var depth = 0
        var quoted = false
        var escaped = false
        var index = start
        while index < payload.endIndex {
            let char = payload[index]
            if escaped {
                escaped = false
            } else if char == "\\" {
                escaped = true
            } else if char == "\"" {
                quoted.toggle()
            } else if !quoted {
                if char == "{" {
                    depth += 1
                } else if char == "}" {
                    depth -= 1
                    if depth == 0 { return index }
                }
            }
            index = payload.index(after: index)
        }
        return nil
    }
}

File: modules/neomovies-core/android/src/main/java/com/neo/neomovies/core/AllohaRuntimeParser.kt
package com.neo.neomovies.core

import org.json.JSONArray
import org.json.JSONObject
import java.net.URI

object AllohaRuntimeParser {

    fun parsePayload(payload: String, baseUrl: String, headers: Map<String, String>): Map<String, Any>? {
        val uri = try { URI(baseUrl) } catch (e: Exception) { return null }

        parseAllohaBNsiStream(payload, uri, headers)?.let { return it }

        firstPreferredStreamURL(payload, uri)?.let { fallback ->
            return mapOf(
                "videoURL" to fallback,
                "audioTracks" to emptyList<Any>(),
                "audioVariants" to emptyList<Any>(),
                "subtitles" to subtitleTracks(payload, uri),
                "qualityVariants" to emptyList<Any>(),
                "httpHeaders" to headers
            )
        }

        return null
    }

    private fun parseAllohaBNsiStream(payload: String, baseUrl: URI, headers: Map<String, String>): Map<String, Any>? {
        val candidates = listOf(payload) + embeddedJSONObjectCandidates(payload)

        for (candidate in candidates) {
            val obj = try { JSONObject(candidate) } catch (e: Exception) { continue }
            val source = try { obj.getJSONArray("hlsSource") } catch (e: Exception) { continue }

            val qualityVariants = mutableListOf<Map<String, Any?>>()
            val audioVariants = mutableListOf<Map<String, Any?>>()
            var masterURL: String? = null

            for (i in 0 until source.length()) {
                val item = try { source.getJSONObject(i) } catch (e: Exception) { continue }
                val quality = try { item.getJSONObject("quality") } catch (e: Exception) { continue }

                val itemVariants = mutableListOf<Map<String, Any?>>()
                var itemMasterURL: String? = null

                for (label in quality.keys()) {
                    val rawValue = quality.get(label)
                    for (rawURL in qualityURLStrings(rawValue)) {
                        val urls = allohaURLs(rawURL, baseUrl)
                        if (masterURL == null) masterURL = urls.firstOrNull { isMasterM3u8(it) }
                        if (itemMasterURL == null) itemMasterURL = urls.firstOrNull { isMasterM3u8(it) }
                        val target = urls.firstOrNull { !isMasterM3u8(it) } ?: urls.firstOrNull() ?: continue

                        val variant: Map<String, Any?> = mapOf(
                            "label" to normalizedQualityLabel(label),
                            "bandwidth" to null,
                            "resolution" to null,
                            "url" to target
                        )
                        itemVariants.add(variant)
                        qualityVariants.add(variant)
                    }
                }

                val chosenURL = itemMasterURL ?: (itemVariants.lastOrNull()?.get("url") as? String)
                if (chosenURL != null) {
                    audioVariants.add(mapOf(
                        "id" to "$i-$chosenURL",
                        "title" to audioVariantTitle(item, i),
                        "url" to chosenURL,
                        "qualityVariants" to itemVariants
                    ))
                }
            }

            val deduped = deduplicatedAudioVariants(audioVariants)
            val pickedURL = (deduped.firstOrNull()?.get("url") as? String)
                ?: masterURL
                ?: (qualityVariants.lastOrNull()?.get("url") as? String)
            pickedURL ?: continue

            return mapOf(
                "videoURL" to pickedURL,
                "audioTracks" to emptyList<Any>(),
                "audioVariants" to deduped,
                "subtitles" to subtitleTracks(payload, baseUrl),
                "qualityVariants" to qualityVariants,
                "httpHeaders" to headers
            )
        }

        return null
    }

    private fun isMasterM3u8(url: String): Boolean {
        val path = url.substringAfterLast("/").substringBefore("?").lowercase()
        return path.contains("master.m3u8")
    }

    private fun deduplicatedAudioVariants(variants: List<Map<String, Any?>>): List<Map<String, Any?>> {
        val seen = mutableSetOf<String>()
        return variants.filter { variant ->
            val key = variant["url"] as? String ?: ""
            seen.add(key)
        }
    }

    private fun audioVariantTitle(item: JSONObject, index: Int): String {
        item.optJSONObject("translation")?.optString("name")
            ?.takeIf { it.isNotEmpty() }?.let { return it }

        for (key in listOf("translationName", "translation_name", "translator", "studio", "voice", "voiceover", "dub", "dubbing", "name", "title", "label")) {
            item.optString(key).takeIf { it.isNotEmpty() && it != "null" }?.let { return it }
        }

        return "Озвучка ${index + 1}"
    }

    private fun qualityURLStrings(value: Any): List<String> {
        return when (value) {
            is String -> value.split(",").map { it.trim() }.filter { it.isNotEmpty() }
            is JSONArray -> (0 until value.length()).flatMap { qualityURLStrings(value.get(it)) }
            else -> emptyList()
        }
    }

    private fun allohaURLs(raw: String, baseUrl: URI): List<String> {
        val decoded = raw.replace("\\/", "/")
        return decoded.split(" or ").mapNotNull { part ->
            val clean = part.trim()
            when {
                clean.startsWith("//") -> "https:$clean"
                clean.startsWith("http://") || clean.startsWith("https://") -> clean
                else -> runCatching { baseUrl.resolve(clean).toString() }.getOrNull()
            }
        }
    }

    private fun normalizedQualityLabel(label: String): String {
        val clean = label.replace("_", " ").trim()
        return if (clean.isEmpty()) "Auto" else clean
    }

    private fun firstPreferredStreamURL(payload: String, baseUrl: URI): String? {
        val pattern = Regex("""https?:\\/\\/[^\"'\s>]+\.(m3u8|mpd)[^\"'\s>]*""", RegexOption.IGNORE_CASE)
        val match = pattern.find(payload) ?: return null
        val value = match.value.replace("\\/", "/")
        return runCatching { baseUrl.resolve(value).toString() }.getOrDefault(value)
    }

    private fun subtitleTracks(payload: String, baseUrl: URI): List<Map<String, String>> {
        val pattern = Regex("""https?:\\/\\/[^\"'\s>]+\.(vtt|srt)[^\"'\s>]*""", RegexOption.IGNORE_CASE)
        return pattern.findAll(payload).mapNotNull { match ->
            val value = match.value.replace("\\/", "/")
            val url = runCatching { baseUrl.resolve(value).toString() }.getOrDefault(value)
            mapOf("url" to url, "label" to "Subtitle", "language" to "ru")
        }.toList()
    }

    private fun embeddedJSONObjectCandidates(payload: String): List<String> {
        val candidates = mutableListOf<String>()
        candidates.addAll(balancedJSONObjectCandidates("\"hlsSource\"", payload))
        candidates.addAll(balancedJSONObjectCandidates("hlsSource", payload))
        return candidates.distinct()
    }

    private fun balancedJSONObjectCandidates(marker: String, payload: String): List<String> {
        val candidates = mutableListOf<String>()
        var searchStart = 0
        while (true) {
            val markerIndex = payload.indexOf(marker, searchStart, ignoreCase = true)
            if (markerIndex < 0) break
            val objectStart = payload.lastIndexOf('{', markerIndex)
            if (objectStart < 0) {
                searchStart = markerIndex + marker.length
                continue
            }
            val objectEnd = balancedObjectEnd(objectStart, payload)
            if (objectEnd < 0) {
                searchStart = markerIndex + marker.length
                continue
            }
            candidates.add(payload.substring(objectStart, objectEnd + 1))
            searchStart = markerIndex + marker.length
        }
        return candidates
    }

    private fun balancedObjectEnd(start: Int, payload: String): Int {
        var depth = 0
        var quoted = false
        var escaped = false
        var i = start
        while (i < payload.length) {
            when {
                escaped -> escaped = false
                payload[i] == '\\' -> escaped = true
                payload[i] == '"' -> quoted = !quoted
                !quoted && payload[i] == '{' -> depth++
                !quoted && payload[i] == '}' -> {
                    depth--
                    if (depth == 0) return i
                }
            }
            i++
        }
        return -1
    }
}
